File: backend/schemas/analytics.py
"""
backend/schemas/analytics.py
=============================
Pydantic schemas for analytics endpoints.
"""

from __future__ import annotations
from pydantic import BaseModel, ConfigDict


class PerformanceSnapshotSchema(BaseModel):
    """Complete performance metrics for a given filter context."""
    account_name: str | None = None
    symbol: str | None = None
    date_from: str | None = None
    date_to: str | None = None
    trade_count: int
    win_count: int
    loss_count: int
    win_rate: float | None
    net_pnl: float
    gross_profit: float
    gross_loss: float
    profit_factor: float | None
    expectancy: float | None
    avg_trade_pnl: float | None
    max_drawdown: float
    sharpe_ratio: float | None
    sortino_ratio: float | None
    calmar_ratio: float | None
    largest_win: float | None
    largest_loss: float | None
    max_consecutive_wins: int
    max_consecutive_losses: int


class TimeBinMetricsSchema(BaseModel):
    """Performance metrics for a single time-bin slot."""
    model_config = ConfigDict(from_attributes=True)

    account_name: str
    symbol: str
    time_slot_ny: str
    trade_count: int
    win_count: int
    loss_count: int
    win_rate: float | None
    total_pnl: float
    avg_pnl: float | None
    expectancy: float | None
    profit_factor: float | None
    sharpe_ratio: float | None
    sortino_ratio: float | None
    max_drawdown: float | None
    wilcoxon_pvalue: float | None
    bh_fdr_pvalue: float | None
    is_promoted: bool


class TimeBinGridSchema(BaseModel):
    """Full time-bin grid with heatmap matrix."""
    resolution: str = "30min"
    timezone: str = "America/New_York"
    alpha: float
    total_slots: int
    promoted_slots: int
    slots: list[TimeBinMetricsSchema]
    heatmap: dict | None = None   # accounts × slots matrix for Plotly


class HeatmapSchema(BaseModel):
    """Plotly-ready heatmap data structure."""
    accounts: list[str]
    slots: list[str]
    matrix: list[list[float | None]]
    promoted_mask: list[list[bool]]
    metric: str


class AccountRankingSchema(BaseModel):
    """Account ranked by a performance metric (Sortino leaderboard)."""
    rank: int
    account_name: str
    symbol: str | None = None
    trade_count: int
    net_pnl: float
    win_rate: float | None
    sharpe_ratio: float | None
    sortino_ratio: float | None
    max_drawdown: float
    promoted_slots: int


class SymbolBreakdownSchema(BaseModel):
    """Per-symbol performance summary."""
    symbol: str
    account_name: str
    trade_count: int
    net_pnl: float
    win_rate: float | None
    sharpe_ratio: float | None
    promoted_slots: int


class PnLCurvePoint(BaseModel):
    date: str
    cumulative_pnl: float
    daily_pnl: float


class DrawdownPoint(BaseModel):
    date: str
    drawdown: float
    drawdown_pct: float

File: frontend/src/types/analytics.ts
/**
 * frontend/src/types/analytics.ts
 */

export interface PerformanceSnapshot {
  account_name: string | null;
  symbol: string | null;
  date_from: string | null;
  date_to: string | null;
  trade_count: number;
  win_count: number;
  loss_count: number;
  win_rate: number | null;
  net_pnl: number;
  gross_profit: number;
  gross_loss: number;
  profit_factor: number | null;
  expectancy: number | null;
  avg_trade_pnl: number | null;
  max_drawdown: number;
  sharpe_ratio: number | null;
  sortino_ratio: number | null;
  calmar_ratio: number | null;
  largest_win: number | null;
  largest_loss: number | null;
  max_consecutive_wins: number;
  max_consecutive_losses: number;
}

export interface TimeBinMetrics {
  account_name: string;
  symbol: string;
  time_slot_ny: string;
  trade_count: number;
  win_count: number;
  loss_count: number;
  win_rate: number | null;
  total_pnl: number;
  avg_pnl: number | null;
  expectancy: number | null;
  profit_factor: number | null;
  sharpe_ratio: number | null;
  sortino_ratio: number | null;
  max_drawdown: number | null;
  wilcoxon_pvalue: number | null;
  bh_fdr_pvalue: number | null;
  is_promoted: boolean;
}

export interface HeatmapData {
  accounts: string[];
  slots: string[];
  matrix: (number | null)[][];
  promoted_mask: boolean[][];
  metric: string;
}

export interface TimeBinGrid {
  resolution: string;
  timezone: string;
  alpha: number;
  total_slots: number;
  promoted_slots: number;
  slots: TimeBinMetrics[];
  heatmap: HeatmapData | null;
}

export interface AccountRanking {
  rank: number;
  account_name: string;
  symbol: string | null;
  trade_count: number;
  net_pnl: number;
  win_rate: number | null;
  sharpe_ratio: number | null;
  sortino_ratio: number | null;
  max_drawdown: number;
  promoted_slots: number;
}

export interface PnLCurvePoint {
  date: string;
  cumulative_pnl: number;
  daily_pnl: number;
}

export interface DrawdownPoint {
  date: string;
  drawdown: number;
  drawdown_pct: number;
}
